cudamapper/src/overlapper_naive.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+* Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#include <cstdio>
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+#include "overlapper_naive.hpp"
+#include "cudamapper/overlapper.hpp"
+#include "matcher.hpp"
+
+
+namespace claragenomics {
+    void OverlapperNaive::print_paf(std::vector<Overlap> overlaps){
+        std::string relative_strand = "+";
+        for(auto overlap: overlaps){
+            std::printf("%s\t%i\t%i\t%i\t%s\t%s\t%i\t%i\t%i\t%i\t%i\t%i\n",
+                    overlap.query_read_name_.c_str(),
+                    0,
+                    overlap.query_start_position_in_read_,
+                    overlap.query_end_position_in_read_,
+                    relative_strand.c_str(),
+                    overlap.target_read_name_.c_str(),
+                    0,
+                    overlap.target_start_position_in_read_,
+                    overlap.target_end_position_in_read_,
+                    overlap.num_residues_,
+                    0,
+                    255
+                    );
+        }
+    }
+
+    std::vector<Overlap> const OverlapperNaive::get_overlaps(const std::vector<Anchor>& anchors, Index& index) {
+
+        std::map<std::pair<int,int>, Overlap> reads_to_overlaps;
```

Comment:
Can this be `std::unordered_map` or is the ordering important?

Reply:
I tried this but it looks like unordered_map is unable to hash the key:

```
[ 75%] Building CXX object cudamapper/CMakeFiles/cudamapper.dir/src/overlapper_naive.cpp.o
In file included from /usr/include/c++/7/bits/hashtable.h:35:0,
                 from /usr/include/c++/7/unordered_map:47,
                 from /home/mike/dev/ClaraGenomicsAnalysis-private/cudamapper/src/overlapper_naive.cpp:14:
/usr/include/c++/7/bits/hashtable_policy.h: In instantiation of ‘struct std::__detail::__is_noexcept_hash<std::pair<int, int>, std::hash<std::pair<int, int> > >’:
/usr/include/c++/7/type_traits:143:12:   required from ‘struct std::__and_<std::__is_fast_hash<std::hash<std::pair<int, int> > >, std::__detail::__is_noexcept_hash<std::pair<int, int>, std::hash<std::pair<int, int> > > >’
/usr/include/c++/7/type_traits:154:31:   required from ‘struct std::__not_<std::__and_<std::__is_fast_hash<std::hash<std::pair<int, int> > >, std::__detail::__is_noexcept_hash<std::pair<int, int>, std::hash<std::pair<int, int> > > > >’
/usr/include/c++/7/bits/unordered_map.h:103:66:   required from ‘class std::unordered_map<std::pair<int, int>, claragenomics::Overlap>’
/home/mike/dev/ClaraGenomicsAnalysis-private/cudamapper/src/overlapper_naive.cpp:24:57:   required from here
/usr/include/c++/7/bits/hashtable_policy.h:87:34: error: no match for call to ‘(const std::hash<std::pair<int, int> >) (const std::pair<int, int>&)’
  noexcept(declval<const _Hash&>()(declval<const _Key&>()))>
           ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/7/bits/move.h:54:0,
                 from /usr/include/c++/7/bits/stl_pair.h:59,
                 from /usr/include/c++/7/bits/stl_algobase.h:64,
                 from /usr/include/c++/7/bits/char_traits.h:39,
                 from /usr/include/c++/7/string:40,
                 from /home/mike/dev/ClaraGenomicsAnalysis-private/cudamapper/src/overlapper_naive.cpp:12:
/usr/include/c++/7/type_traits: In instantiation of ‘struct std::__not_<std::__and_<std::__is_fast_hash<std::hash<std::pair<int, int> > >, std::__detail::__is_noexcept_hash<std::pair<int, int>, std::hash<std::pair<int, int> > > > >’:
/usr/include/c++/7/bits/unordered_map.h:103:66:   required from ‘class std::unordered_map<std::pair<int, int>, claragenomics::Overlap>’
/home/mike/dev/ClaraGenomicsAnalysis-private/cudamapper/src/overlapper_naive.cpp:24:57:   required from here
/usr/include/c++/7/type_traits:154:31: error: ‘value’ is not a member of ‘std::__and_<std::__is_fast_hash<std::hash<std::pair<int, int> > >, std::__detail::__is_noexcept_hash<std::pair<int, int>, std::hash<std::pair<int, int> > > >’
     : public __bool_constant<!bool(_Pp::value)>
                               ^~~~~~~~~~~~~~~~
```

Reply:
True. We can leave it as `map` then. Let's see if it becomes a performance bottleneck...